manticore/native/state.py

```diff
@@ -55,56 +66,109 @@ def __enter__(self) -> "State":
             - set of hooks for specified after or before
             - string of callback event
             - State function that handles the callback
         """
-        return (
-            (self._hooks, "will_execute_instruction", self._state_hook_callback)
-            if not after
-            else (self._after_hooks, "did_execute_instruction", self._state_after_hook_callback)
-        )
-
-    def remove_hook(self, pc: Optional[int], callback: HookCallback, after: bool = False) -> bool:
+        if not syscall:
+            return (
+                (self._hooks, "will_execute_instruction", self._state_hook_callback)
+                if not after
+                else (self._after_hooks, "did_execute_instruction", self._state_after_hook_callback)
+            )
+        else:
+            return (
+                (self._sys_hooks, "will_invoke_syscall", self._state_sys_hook_callback)
+                if not after
+                else (
+                    self._sys_after_hooks,
+                    "did_invoke_syscall",
+                    self._state_sys_after_hook_callback,
+                )
+            )
+
+    def remove_hook(
+        self,
+        pc_or_sys: Optional[Union[int, str]],
+        callback: HookCallback,
+        after: bool = False,
+        syscall: bool = False,
+    ) -> bool:
         """
         Remove a callback with the specified properties
-        :param pc: Address of instruction to remove from
-        :param callback: The callback function that was at the address
+        :param pc_or_sys: Address of instruction, syscall number, or syscall name to remove hook from
+        :param callback: The callback function that was at the address (or syscall)
         :param after: Whether it was after instruction executed or not
+        :param syscall: Catch a syscall invocation instead of instruction?
         :return: Whether it was removed
         """
-        hooks, when, _ = self._get_hook_context(after)
-        cbs = hooks.get(pc, set())
+
+        if isinstance(pc_or_sys, str):
+            table = getattr(linux_syscalls, self._platform.current.machine)
+            for index, name in table.items():
+                if name == pc_or_sys:
+                    pc_or_sys = index
+                    break
+            if isinstance(pc_or_sys, str):
+                logger.warning(
+                    f"{pc_or_sys} is not a valid syscall name in architecture {self._platform.current.machine}. "
+                    "Please refer to manticore/platforms/linux_syscalls.py to find the correct name."
+                )
+                return False
+
+        hooks, when, _ = self._get_hook_context(after, syscall)
+        cbs = hooks.get(pc_or_sys, set())
         if callback in cbs:
             cbs.remove(callback)
         else:
             return False
 
-        if len(hooks.get(pc, set())) == 0:
-            del hooks[pc]
+        if len(hooks.get(pc_or_sys, set())) == 0:
```

Comment:
Extremely minor nit: for collections that support it, Python generally prefers `if not x` over `if len(x) == 0`